src/lightning/data/cache/reader.py

```diff
@@ -0,0 +1,186 @@
+
+                self._chunks.extend(data["chunks"])
+
+        self._config["data_spec"] = treespec_loads(self._config["data_spec"])
+
+        for chunk in self._chunks:
+            start, end = chunk["interval"]
+            if (end - start) != chunk["samples"]:
+                raise Exception(
+                    "The config intervals doesn't match the number of samples. This shouldn't have happened."
+                )
+            self._intervals.append(chunk["interval"])
+
+        self._length = sum([chunk["samples"] for chunk in self._chunks])
+
+    @property
+    def intervals(self):
+        return self._intervals
+
+    @property
+    def data_format(self):
+        return self._config["data_format"]
+
+    @property
+    def config(self):
+        return self._config
+
+    def __getitem__(self, index: int) -> Tuple[str, int, int]:
+        """Find the associated chunk metadata."""
+        for interval_config, internal in enumerate(self._intervals):
+            if internal[0] <= index and index < internal[1]:
+                chunk = self._chunks[interval_config]
+                mapping = chunk["mapping"][str(index)]
+                return os.path.join(self._cache_dir, chunk["filename"]), *mapping
+        raise Exception(f"The chunk interval weren't properly defined. Found {self._intervals} for index {index}.")
+
+    @classmethod
+    def load(cls, cache_dir: str) -> Optional["ChunksConfig"]:
+        files = os.listdir(cache_dir)
+        index_filenames = sorted([f for f in files if f.endswith("index.json")])
+        if not index_filenames:
+            return None
+        return ChunksConfig(cache_dir, index_filenames)
+
+    def __len__(self) -> int:
+        return self._length
+
+
+class BinaryReader:
+    def __init__(self, cache_dir: str, compression: Optional[str] = None):
+        """The BinaryReader enables to read chunked dataset in an efficient way.
+
+        Arguments:
+            cache_dir: The path to cache folder
+            compression: The algorithm to decompress the chunks.
+
+        """
+
+        super().__init__()
+        self._cache_dir = cache_dir
+
+        if not os.path.exists(self._cache_dir):
+            raise FileNotFoundError(f"The provided cache_dir `{self._cache_dir}` doesn't exist.")
+
+        self._compression = compression
+        self._config = None
+        self._intervals = None
+
+        self._chunks_data = {}
+        self._serializers: Dict[str, Serializer] = _SERIALIZERS
+
+        self._env = _DistributedEnv.detect()
+        self._config: Optional[ChunksConfig] = None
+
+    def _try_load_config(self):
+        """Try to load the chunks config if the index files are available."""
+        self._config = ChunksConfig.load(self._cache_dir)
+
+    def read(self, index: int):
+        """Read an item for the given from a chunk.
+
+        If the chunk isn't available, it will be downloaded.
```

Comment:
Since we own the sampler and the data loader, couldn't we avoid realizing we need a chunk at the last minute and instead prefetch the chunks we need, as a future optimization?

Reply:
Yes, I built the code to make this dead simple. 
This would be done in the next PR. 
There would be an executor to download the chunks. We will measure to time to download, time to batch in the readers and infer from the current index when to load the next chunk.
Plus, we would have an eviction logic if the dataset is larger than the current machine size.

Reply:
Added a comment.